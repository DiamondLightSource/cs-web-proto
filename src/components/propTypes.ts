import PropTypes from "prop-types";

export const StringProp = PropTypes.string.isRequired;
export const StringPropOpt = PropTypes.string;

export const IntProp = PropTypes.number.isRequired;
export const IntPropOpt = PropTypes.number;

export const FloatProp = PropTypes.number.isRequired;
export const FloatPropOpt = PropTypes.number;

export const BoolProp = PropTypes.bool.isRequired;
export const BoolPropOpt = PropTypes.bool;

export const PvProp = PropTypes.string.isRequired;
export const PvPropOpt = PropTypes.string;

export const StringOrNumPropOpt = PropTypes.oneOfType([
  PropTypes.string,
  PropTypes.number
]);
export const StringOrNumProp = StringOrNumPropOpt.isRequired;

// Number of prop types organised into useable sections to form more
// complex units
export const ContainerFeaturesPropType = {
  margin: StringPropOpt,
  padding: StringPropOpt,
  border: StringPropOpt,
  minWidth: StringPropOpt,
  maxWidth: StringPropOpt
};

const OpenWebpagePropType = PropTypes.shape({
  type: StringProp,
  url: StringProp,
  description: StringPropOpt
});

const OpenPagePropType = PropTypes.shape({
  type: StringProp,
  page: StringProp,
  location: StringProp,
  macros: StringProp,
  description: StringPropOpt
});

const ClosePagePropType = PropTypes.shape({
  type: StringProp,
  location: StringProp,
  description: StringPropOpt
});

const WritePvPropType = PropTypes.shape({
  type: StringProp,
  pvName: StringProp,
  value: PropTypes.oneOfType([StringPropOpt, FloatPropOpt]),
  description: StringPropOpt
});

const ActionPropType = PropTypes.oneOfType([
  OpenPagePropType,
  ClosePagePropType,
  WritePvPropType,
  OpenWebpagePropType
]);

export const ActionsPropType = PropTypes.shape({
  executeAsOne: BoolPropOpt,
  actions: PropTypes.arrayOf(ActionPropType).isRequired
});
