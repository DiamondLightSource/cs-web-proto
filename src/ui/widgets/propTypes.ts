import PropTypes, { InferProps } from "prop-types";

export type ExcludeNulls<T> = {
  [P in keyof T]: Exclude<T[P], null>;
};
export type InferWidgetProps<T> = ExcludeNulls<InferProps<T>>;

export const StringProp = PropTypes.string.isRequired;
export const StringPropOpt = PropTypes.string;

export const IntProp = PropTypes.number.isRequired;
export const IntPropOpt = PropTypes.number;

export const FloatProp = PropTypes.number.isRequired;
export const FloatPropOpt = PropTypes.number;

export const BoolProp = PropTypes.bool.isRequired;
export const BoolPropOpt = PropTypes.bool;

export const PvProp = PropTypes.string.isRequired;
export const PvPropOpt = PropTypes.string;

export const ChildrenProp = PropTypes.node.isRequired;
export const ChildrenPropOpt = PropTypes.node;

export const ObjectProp = PropTypes.object.isRequired;
export const ObjectPropOpt = PropTypes.object;

export const MacrosPropOpt = PropTypes.objectOf(PropTypes.string.isRequired);
export const MacrosProp = PropTypes.objectOf(PropTypes.string.isRequired)
  .isRequired;

export const StringOrNumPropOpt = PropTypes.oneOfType([
  PropTypes.string,
  PropTypes.number
]);
export const StringOrNumProp = StringOrNumPropOpt.isRequired;

export function ChoicePropOpt(
  options: string[]
): PropTypes.Requireable<string> {
  return PropTypes.oneOf(options);
}
export function ChoiceProp(options: string[]): PropTypes.Validator<string> {
  return PropTypes.oneOf(options).isRequired;
}

// Number of prop types organised into useable sections to form more
// complex units
export const ContainerFeaturesPropType = {
  margin: StringPropOpt,
  padding: StringPropOpt,
  border: StringPropOpt,
  minWidth: StringPropOpt,
  maxWidth: StringPropOpt
};

const OpenWebpagePropType = PropTypes.shape({
  type: StringProp,
  openWebpageInfo: PropTypes.shape({
    url: StringProp,
    description: StringPropOpt
  }).isRequired
});

const OpenPagePropType = PropTypes.shape({
  type: StringProp,
  openPageInfo: PropTypes.shape({
    page: StringProp,
    location: StringProp,
    macros: StringProp,
    description: StringPropOpt
  }).isRequired
});

const ClosePagePropType = PropTypes.shape({
  type: StringProp,
  closePageInfo: PropTypes.shape({
    location: StringProp,
    description: StringPropOpt
  }).isRequired
});

const WritePvPropType = PropTypes.shape({
  type: StringProp,
  writePvInfo: PropTypes.shape({
    pvName: StringProp,
    value: PropTypes.oneOfType([StringPropOpt, FloatPropOpt]),
    description: StringPropOpt
  }).isRequired
});

const ActionPropType = PropTypes.oneOfType([
  OpenPagePropType,
  ClosePagePropType,
  WritePvPropType,
  OpenWebpagePropType
]);

export const ActionsPropType = PropTypes.shape({
  executeAsOne: BoolPropOpt,
  actions: PropTypes.arrayOf(ActionPropType).isRequired
});
