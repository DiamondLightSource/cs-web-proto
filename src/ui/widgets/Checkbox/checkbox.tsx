import React from "react";
import { Widget } from "../widget";
import { InferWidgetProps } from "../propTypes";
import { PVComponent, PVWidgetPropType } from "../widgetProps";
import { registerWidget } from "../register";

export const CheckboxProps = {};

export type CheckboxComponentProps = InferWidgetProps<typeof CheckboxProps> &
  PVComponent;

export const CheckboxComponent = (
  props: CheckboxComponentProps
): JSX.Element => {
  return <div>{"Component is rendered"}</div>;
};

const LedWidgetProps = {
  ...CheckboxProps,
  ...PVWidgetPropType
};

export const Checkbox = (
  props: InferWidgetProps<typeof LedWidgetProps>
): JSX.Element => <Widget baseWidget={CheckboxComponent} {...props} />;

registerWidget(Checkbox, LedWidgetProps, "checkbox");
