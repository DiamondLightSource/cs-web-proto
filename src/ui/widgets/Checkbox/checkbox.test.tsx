import React from "react";
import { CheckboxComponent, CheckboxProps } from "./checkbox";
import renderer, { ReactTestRenderer } from "react-test-renderer";
import { render, fireEvent, screen } from "@testing-library/react";
import { InferWidgetProps } from "../propTypes";
import FormControlLabel from "@material-ui/core/FormControlLabel";
import { Checkbox as MaterialCheckbox } from "@material-ui/core";
import { ensureWidgetsRegistered } from "..";
ensureWidgetsRegistered();

const checkboxRenderer = (checkboxProps: any): ReactTestRenderer => {
  return renderer.create(
    <CheckboxComponent {...checkboxProps} readonly={true} />
  );
};

type CheckboxComponentProps = InferWidgetProps<typeof CheckboxProps>;

describe("<CheckboxComponent />", (): void => {
  it("default properties are applied", (): void => {
    const testRenderer = checkboxRenderer({});

    const form = testRenderer.root.findByType(FormControlLabel);

    expect(form.props.style).toEqual({ width: "10px", height: "10px" });

    const checkbox = testRenderer.root.findByType(MaterialCheckbox);

    expect(checkbox.props.checked).toBe(false);
    expect(checkbox.props.disableRipple).toBe(true);
  });

  it("height and width are applied to style", (): void => {
    const checkboxProps: CheckboxComponentProps = {
      width: 15,
      height: 20
    };

    const testRenderer = checkboxRenderer(checkboxProps);

    const form = testRenderer.root.findByType(FormControlLabel);

    const formStyle = form.props.style;

    expect(formStyle.width).toBe("15px");
    expect(formStyle.height).toBe("20px");
  });

  it("matches snapshot", (): void => {
    const checkboxProps: CheckboxComponentProps = {
      label: "Here is a label",
      width: 15,
      height: 20
    };
    const { asFragment } = render(
      <CheckboxComponent {...checkboxProps} readonly={true} connected={true} />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("doesn't create a label if there is no text", (): void => {
    const checkboxProps: CheckboxComponentProps = {};

    render(
      <CheckboxComponent {...checkboxProps} readonly={true} connected={true} />
    );

    expect((): void => {
      screen.getByRole("checkbox", {
        name: /Here is a label/i
      });
    }).toThrowErrorMatchingSnapshot();
  });

  it("creates a label when there is text", (): void => {
    const checkboxProps: CheckboxComponentProps = {
      label: "Here is a label",
      width: 15,
      height: 20
    };

    render(
      <CheckboxComponent {...checkboxProps} readonly={true} connected={true} />
    );

    expect(
      screen.getByRole("checkbox", {
        name: /Here is a label/i
      })
    ).toBeInTheDocument();
  });

  describe("user input changes properties", (): void => {
    it("clicking on checkbox toggles it", (): void => {
      const checkboxProps: CheckboxComponentProps = {
        label: "Here is a label",
        width: 15,
        height: 20
      };

      render(
        <CheckboxComponent
          {...checkboxProps}
          connected={true}
          readonly={true}
        />
      );

      const items: any = screen.getByRole("checkbox", {
        name: /Here is a label/i
      });

      expect(items.checked).toBe(false);
      fireEvent.click(items);
      expect(items.checked).toBe(true);
    });
  });
});
