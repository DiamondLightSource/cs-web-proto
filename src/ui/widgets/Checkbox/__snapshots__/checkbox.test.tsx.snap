// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckboxComponent /> doesn't create a label if there is no text 1`] = `
"Unable to find an accessible element with the role \\"checkbox\\" and name \`/Here is a label/i\`

Here are the accessible roles:

  checkbox:

  Name \\"\\":
  [36m<input[39m
    [33mclass[39m=[32m\\"PrivateSwitchBase-input-6\\"[39m
    [33mdata-indeterminate[39m=[32m\\"false\\"[39m
    [33mtype[39m=[32m\\"checkbox\\"[39m
    [33mvalue[39m=[32m\\"\\"[39m
  [36m/>[39m

  --------------------------------------------------

[36m<body>[39m
  [36m<div>[39m
    [36m<div>[39m
      [36m<label[39m
        [33mclass[39m=[32m\\"MuiFormControlLabel-root\\"[39m
        [33mstyle[39m=[32m\\"width: 10px; height: 10px;\\"[39m
      [36m>[39m
        [36m<span[39m
          [33maria-disabled[39m=[32m\\"false\\"[39m
          [33mclass[39m=[32m\\"MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-3 MuiCheckbox-root makeStyles-root-1 MuiCheckbox-colorSecondary MuiIconButton-colorSecondary\\"[39m
        [36m>[39m
          [36m<span[39m
            [33mclass[39m=[32m\\"MuiIconButton-label\\"[39m
          [36m>[39m
            [36m<input[39m
              [33mclass[39m=[32m\\"PrivateSwitchBase-input-6\\"[39m
              [33mdata-indeterminate[39m=[32m\\"false\\"[39m
              [33mtype[39m=[32m\\"checkbox\\"[39m
              [33mvalue[39m=[32m\\"\\"[39m
            [36m/>[39m
            [36m<svg[39m
              [33maria-hidden[39m=[32m\\"true\\"[39m
              [33mclass[39m=[32m\\"MuiSvgIcon-root\\"[39m
              [33mfocusable[39m=[32m\\"false\\"[39m
              [33mviewBox[39m=[32m\\"0 0 24 24\\"[39m
            [36m>[39m
              [36m<path[39m
                [33md[39m=[32m\\"M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z\\"[39m
              [36m/>[39m
            [36m</svg>[39m
          [36m</span>[39m
        [36m</span>[39m
        [36m<span[39m
          [33mclass[39m=[32m\\"MuiTypography-root MuiFormControlLabel-label MuiTypography-body1\\"[39m
        [36m/>[39m
      [36m</label>[39m
    [36m</div>[39m
  [36m</div>[39m
[36m</body>[39m"
`;

exports[`<CheckboxComponent /> matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <label
      class="MuiFormControlLabel-root"
      style="width: 15px; height: 20px;"
    >
      <span
        aria-disabled="false"
        class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-3 MuiCheckbox-root makeStyles-root-1 MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
      >
        <span
          class="MuiIconButton-label"
        >
          <input
            class="PrivateSwitchBase-input-6"
            data-indeterminate="false"
            type="checkbox"
            value=""
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
        </span>
      </span>
      <span
        class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
      >
        <div
          class="Label Label"
          style="justify-content: center;"
        >
          Here is a label
        </div>
      </span>
    </label>
  </div>
</DocumentFragment>
`;
